zookeeper-api/src/main/java/org/apache/helix/zookeeper/datamodel/serializer/ZNRecordStreamingSerializer.java

```diff
@@ -154,20 +155,31 @@ private static int getListFieldBound(ZNRecord record) {
       g.close();
       serializedBytes = baos.toByteArray();
       // apply compression if needed
-      if (record.getBooleanField("enableCompression", false) || serializedBytes.length > ZNRecord.SIZE_LIMIT) {
+      if (ZNRecordUtil.shouldCompress(record, serializedBytes.length)) {
         serializedBytes = GZipCompressionUtil.compress(serializedBytes);
       }
     } catch (Exception e) {
-      LOG.error("Exception during data serialization. Will not write to zk. Data (first 1k): "
-          + new String(baos.toByteArray()).substring(0, 1024), e);
+      if (serializedBytes.length == 0 || GZipCompressionUtil.isCompressed(serializedBytes)) {
+        serializedBytes = baos.toByteArray();
+      }
+      int firstBytesLength = Math.min(serializedBytes.length, 1024);
+      // TODO: remove logging first N bytes of data to reduce log size.
+      LOG.error("Exception during data serialization. Will not write to zk."
+              + " The first {} bytes of data: {}", firstBytesLength,
+          new String(serializedBytes, 0, firstBytesLength), e);
       throw new ZkClientException(e);
     }
     // check size
-    if (serializedBytes.length > ZNRecord.SIZE_LIMIT) {
-      LOG.error("Data size larger than 1M, ZNRecord.id: " + record.getId()
-          + ". Will not write to zk. Data (first 1k): "
-          + new String(serializedBytes).substring(0, 1024));
-      throw new ZkClientException("Data size larger than 1M, ZNRecord.id: " + record.getId());
+    int compressThreshold = ZNRecordUtil.getCompressThreshold();
+    if (serializedBytes.length > compressThreshold) {
+      int firstBytesLength = Math.min(serializedBytes.length, 1024);
+      // TODO: remove logging first N bytes of data to reduce log size.
+      LOG.error("Data size: {} is greater than {} bytes, ZNRecord.id: {}."
+              + " Data will not be written to Zookeeper. The first {} bytes of data: {}",
+          serializedBytes.length, compressThreshold, record.getId(), firstBytesLength,
+          new String(serializedBytes, 0, firstBytesLength));
+      throw new ZkClientException("Data size: " + serializedBytes.length + " is greater than "
+          + compressThreshold + " bytes, ZNRecord.id: " + record.getId());
```

Comment:
I think it might be a good idea to also log and include in the exception message whether the ZNRecord was compressed or not. Adding a boolean `isCompressed` would do?